ocaml/xapi/agility.ml

```diff
@@ -107,48 +107,63 @@ module NetworkSet = Set.Make(struct type t = API.ref_network let compare = compa
 
 let empty_cache = (SRSet.empty, NetworkSet.empty)
 
+let (>>=) = Either.Monad.bind
+
 let caching_vm_res_assert_agile ~__context (ok_srs, ok_networks) vm_res =
 	(* Any kind of vGPU means that the VM is not agile. *)
-	if vm_res.VMResources.vm_rec.API.vM_VGPUs <> [] then begin
-		let vm_string =
-			match vm_res.VMResources.status with
-			| `Local vm_ref -> Ref.string_of vm_ref
-			| `Incoming _ -> Ref.string_of Ref.null
-		in
-		raise (Api_errors.Server_error
-			(Api_errors.vm_has_vgpu, [vm_string]))
-	end;
+	let check_vm_rec vm_rec =
+		if vm_rec.API.vM_VGPUs <> []
+		then Either.Left `Has_vgpu
+		else Either.Right ()
+	in
 	(* All referenced SRs should be shared. *)
 	let check_sr ok_srs sr =
 		if SRSet.mem sr ok_srs
-		then ok_srs
+		then Either.Right ok_srs
 		else
-			if not (is_sr_properly_shared ~__context ~self:sr)
-			then raise (Api_errors.(Server_error(ha_constraint_violation_sr_not_shared, [Ref.string_of sr])))
-			else SRSet.add sr ok_srs in
+			if is_sr_properly_shared ~__context ~self:sr
+			then Either.Right (SRSet.add sr ok_srs)
```

Comment:
Resurrecting @jonludlam's comment:

> Is it expensive to call 'is_sr_properly_shared'? We're going to a lot of trouble to thread through the cache of ok_srs and networks through this function. Why not evaluate it up front for all SRs/networks? 

To which I said:

> The reason this caching was done originally was to reduce the work the HA planner would have to do. If we evaluate is_sr_properly_shared for all SRs up front, I worry we're going to be running it a lot for all the local SRs on each host, when in a typical HA setup all we really care about is the shared SRs, since that's where the protected VMs are going to be running.